packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol

```diff
@@ -997,4 +1381,68 @@ contract ConstantFlowAgreementV1 is
         // divisor cannot be zero with existing outflow
         return totalRewardLeft / totalCFAOutFlowrate > int256(liquidationPeriod - patricianPeriod);
     }
+
+    /**************************************************************************
+     * ACL Pure Functions
+     *************************************************************************/
+
+    function _generateFlowOperatorId(address sender, address flowOperator) private pure returns(bytes32 id) {
+        return keccak256(abi.encode("flowOperator", sender, flowOperator));
+    }
+
+    //
+    // Data packing:
+    //
+    // WORD A: | reserved  | permissions | reserved | flowRateAllowance |
+    //         | 120       | 8           | 32       | 96                |
+    //
+    // NOTE:
+    // - flowRateAllowance has 96 bits length
+    // - permissions is an 8-bit octo bitmask
+    // - ...0 0 0 (...delete update create)
+
+    function _encodeFlowOperatorData
+    (
+        FlowOperatorData memory flowOperatorData
+    )
+        internal pure
+        returns(bytes32[] memory data)
+    {
+        data = new bytes32[](1);
+        data[0] = bytes32(
+            uint256(flowOperatorData.permissions) << 128 |
+            uint256(int256(flowOperatorData.flowRateAllowance))
```

Comment:
This is potentially dangerous.
When converting negative integers to an unsigned integer, Solidity maps that value to `max_value_of_type - abs(input_value)`. E,g. `uint16(-1)` becomes 65535.  
This means that, against the intuition, you can't rely on the upper 128 bits being 0 after the type conversion to uint256.
Hence, a flowRateAllowance of -1 flips all bits to 1, incl. the permissions bits.

![image](https://user-images.githubusercontent.com/5479136/161826370-40675100-a608-460f-93c0-0be0cf300ce4.png)

In this case, nothing bad may happen because the msg.sender has no incentive to hack themselves this way. Still, `updateFlowOperatorPermissions` doesn't prevent setting a negative flowRateAllowance, allowing for undefined states (reserved bits set). Also, a msg.sender abusing this could set permissions flags which aren't reflected in the emitted event.

Assuming that we'll never want negative values for the flowRateAllowance, it would probably be best to make it an uint in `FlowOperatorData`.

Reply:
suggested to have a input parameter check of flowRateAllowance to be >= 0.

And here we could use assert for invariance check.